### PR TITLE
Generate a docs table for file format examples

### LINKED ISSUE #1489: File Format List
We have the start of a list of supported file formats at https://girder.github.io/large_image/formats.html.  It would be great to expand this into a table with <extension> <format short name and link> <wsi (bool)> <geo (bool)> <associated images (bool)> <variants and sample file(s) link if known>.  CI could validate that all those links work, too.

That is, something like

| Extension | Format | Tiled | Geospatial | Associated Images | Variants and Sample Files | Source Readers | Write |
| --- | --- | --- | --- | --- | --- | --- | --- |
| tiff | Tagged Image File Format | optional | optional | yes | ... so many ... | tiff, ometiff, tifffile, gdal, mapnik, rasterio, bioformats, vips, openslide | yes |

### PR DESCRIPTION
**Pull request by annehaley**
Resolves #1489 

- [x] Create a new pooch datastore to fetch format example files
- [x] Write a function to open format examples and evaluate table columns
- [x] Write a function to save results to disk as an RST-formatted table
- [x] Invoke the file fetching and table generation as a step in `mkdocs.sh`
- [x] Copy `parakon` hosted files to a more accessible location
- [x] Confirm that all format examples are public, small enough to download during CI, multiframe if possible, and visually interesting

### CODE REVIEW

File: docs/formats.rst [hunk 1]
```diff
@@ -6,7 +6,15 @@ Preferred Extensions and Mime Types
...
 
-The file extensions and mime types that are listed by the core sources that can affect source processing order are listed below.  See ``large_image.listSources()`` for details about priority of the different source and the ``large_image.constants.SourcePriority`` for the priority meaning.
+The file extensions and mime types that are listed by the core sources that can affect source processing order are listed below.  See ``large_image.listSources()`` for details about priority of the different sources and the ``large_image.constants.SourcePriority`` for the priority meaning.
+
+The following table describes the primary formats supported by ``large-image`` and the acceptable extensions for each format. This table also describes the ``large-image-source-*`` modules that can be used to read each format.
```
Comment (manthey) [inline docs/formats.rst:11]: I'd reword this a little.  I'd change "primary" to "some of the" and replace "acceptable" with "common".

My rationale is that we list formats I wouldn't consider primary (GIF and BMP, even LIF).  And, we actually can work with files regardless of extension, and some have multiple common extensions (e.g., .tif, .tiff; .jpeg, .jpg (and even .jpe); .j2k, .jp2, etc.).

File: docs/generate_format_table.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+                                    name=name,
+                                    reference=reference,
+                                    extension=extension,
+                                    filename=filename,
+                                    url=url,
+                                    tilesource=tilesource_name,
+                                    multiframe=(
+                                        False if tilesource_name in NO_MULTIFRAME_SOURCES else
+                                        True if s.getMetadata().get('frames') is not None else
+                                        'Maybe; no multiframe sample found.'
```
Comment (manthey) [inline docs/generate_format_table.py:41]: I'm wondering if we should just leave the maybes blank instead of saying maybe.

File: docs/formats.rst [hunk 3]
```diff
@@ -6,7 +6,15 @@ Preferred Extensions and Mime Types
...
+
+For a full list of all accepted extensions, see :ref:`extensions`.
```
Comment (manthey) [inline docs/formats.rst:13]: I'd change this to "For a list of known extensions, see ..."; we accept others, just they aren't explicitly enumerated by any source.

File: docs/format_examples_datastore.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
+import pooch
+from pathlib import Path
+
+EXAMPLES_FOLDER = Path('format_examples')
+
+format_examples = [
+    dict(
```
Comment (manthey) [inline docs/format_examples_datastore.py:7]: From a priority order, this should either present alphabetically OR we should list what we thing are the most common first (maybe tiff / svs / ome-tiff / ndpi / zarr / nd2 / czi / dicom / (others)).

File: docs/format_examples_datastore.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+        name='TIFF (Tagged Image File Format)',
```
Comment (manthey) [inline docs/format_examples_datastore.py:8]: In the actual table presentable, can we make the explanation of the acronyms less prominent?  Lots of people know JPEG, PNG, DICOM, etc. and never know what they stand for; having the primary name stand out more would be better.

File: docs/format_examples_datastore.py [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+            ),
+            dict(
+                filename='clouds.jpeg',
+                url='https://data.kitware.com/api/v1/file/5afd936c8d777f15ebe1b4d4/download',
+                hash='fb195bb9dae13aa85d9f46f86ab9441a3b436dabbf57c68149453cba691d4391',
+            ),
+        ],
+    ),
+    dict(
+        name='JPEG (Joint Photographic Experts Group) 2000',
```
Comment (manthey) [inline docs/format_examples_datastore.py:66]: I would reorder this "JPEG 2000 (Joint ...)", rather than explain the acronym in the middle of the format name.

File: docs/format_examples_datastore.py [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+        examples=[
+            dict(
+                filename='erdas_foo.jp2',
+                url='https://github.com/OSGeo/gdal/raw/master/autotest/gdrivers/data/jpeg2000/erdas_foo.jp2',
+                hash='d08653919c41a49cd31f9116278fe6a6d8c189dc6379a280810f65df11bde006',
+            ),
+        ],
+    ),
+    dict(
+        name='GIF (Graphics Interchange Format)',
```
Comment (manthey) [inline docs/format_examples_datastore.py:77]: Pedantically wasn't this originally "CompuServe Graphics Interchange Format"?  (don't change it)

File: docs/generate_format_table.py [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+                for tilesource_name, readable in large_image.canReadList(filepath):
+                    tilesource = available_tilesources.get(tilesource_name)
+                    if readable and tilesource:
+                        try:
+                            s = tilesource(filepath)
+                            results.append(
+                                dict(
+                                    name=name,
+                                    reference=reference,
+                                    extension=extension,
```
Comment (manthey) [inline docs/generate_format_table.py:34]: When there are multiple common extensions, should we list them here?  (e.g. tif and tiff, jp2 and j2k, etc)?

**Review by manthey — COMMENTED** [review]
Overall this is great.

File: docs/formats.rst (hunk 1 above)
Comment (annehaley) [reply docs/formats.rst]: Updated in [ba08389](https://github.com/girder/large_image/pull/1622/commits/ba08389bb4f16b06377ebbf3bb2d78e1b235416d)

File: docs/generate_format_table.py (hunk 2 above)
Comment (annehaley) [reply docs/generate_format_table.py]: Changed to "Unknown" per @jeffbaumes' suggestion in [2f65780](https://github.com/girder/large_image/pull/1622/commits/2f65780fb164e0198781de5adfafa03dd033b0d6)

File: docs/formats.rst (hunk 3 above)
Comment (annehaley) [reply docs/formats.rst]: Changed in [bd520a3](https://github.com/girder/large_image/pull/1622/commits/bd520a302dacfed5d273c9dc3af89d86e0723980)

File: docs/format_examples_datastore.py (hunk 4 above)
Comment (annehaley) [reply docs/format_examples_datastore.py:7]: Reordered in [22b5b50](https://github.com/girder/large_image/pull/1622/commits/22b5b5050f8768ad27e724693d2adc4d92955d53)

File: docs/format_examples_datastore.py (hunk 5 above)
Comment (annehaley) [reply docs/format_examples_datastore.py]: Long names moved to tooltips in [544cc37](https://github.com/girder/large_image/pull/1622/commits/544cc37c7745974555ebe2d01b325cbe7d43e9d8)

File: docs/generate_format_table.py (hunk 8 above)
Comment (annehaley) [reply docs/generate_format_table.py]: Changed extension column to show a list of possible extensions for the type (instead of only getting the extension from the example file): [f11383b](https://github.com/girder/large_image/pull/1622/commits/f11383b91889ec1b1a95b1495d6b704e6b9f0d4a)

**Comment by manthey** [comment]
Do you think it would be possible to make the format column have both a link and a tooltip using custom html?  That is
```
.. raw:: html

   <a href="{reference_link}" title={long_name}>{col_value}</a>
```
should work (embedded in the rest of the table cell appropriately).

**Review by manthey — APPROVED** [review]
I'd love to figure out the tooltip-and-link rst so we can not have the word (spec) repeated in each row, but that could be a different PR.

I also wonder if we want to find or generate more relevant (e.g., either WSI or geospatial) samples for some of the download examples, but that, too, can be a different PR.

And, I note that a minor change in the PIL source will let it show GIF as multi-frame (the first image was palette mode which meant that we decided the image stack wasn't consistent enough to show the frames; see a different PR for handling this).
